src/Microsoft.Azure.SignalR/ClientInvocation/RoutedClientResultsManager.cs

```diff
@@ -0,0 +1,86 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Diagnostics;
+using System.Threading.Tasks;
+using System.Threading;
+using Microsoft.AspNetCore.SignalR.Protocol;
+using System.Collections.Generic;
+using Microsoft.Azure.SignalR.Protocol;
+
+namespace Microsoft.Azure.SignalR
+{
+    internal class RoutedClientResultsManager: IRoutedClientResultsManager
+    {
+        private readonly ConcurrentDictionary<string,  RoutedInvocation> _routedInvocations = new();
+        private readonly ConcurrentDictionary<string, List<string>> _serviceMapping = new();
+
+        public void AddRoutedInvocation(string connectionId, string invocationId, string callerServerId, string instanceId, CancellationToken cancellationToken)
+        {
+            cancellationToken.Register(() => TryCompleteResult(connectionId, CompletionMessage.WithError(invocationId, "Canceled")));
+
+            var result = _routedInvocations.TryAdd(invocationId, new RoutedInvocation(connectionId, callerServerId));
+            Debug.Assert(result);
+
+            AddServiceMappingMessage(instanceId, invocationId);
+        }
+
+        public bool TryCompleteResult(string connectionId, CompletionMessage message)
+        {
+            if (_routedInvocations.TryGetValue(message.InvocationId, out var item))
+            {
+                if (item.ConnectionId != connectionId)
+                {
+                    throw new InvalidOperationException($"Connection ID '{connectionId}' is not valid for invocation ID '{message.InvocationId}'.");
+                }
+                return _routedInvocations.TryRemove(message.InvocationId!, out _);
+            }
+            else
+            {
+                // connection was disconnected or someone else completed the invocation
+                return false;
+            }
+        }
+
+        public bool CheckRoutedInvocation(string invocationId)
+        {
+            return _routedInvocations.TryGetValue(invocationId, out _);
+        }
+
+        public bool TryGetInvocationReturnType(string invocationId, out Type type)
+        {
+            if (_routedInvocations.TryGetValue(invocationId, out _))
+            {
+                type = typeof(object);
```

Comment:
Are you sure you want to use `object`? If the deserialized object is used by different protocols you'll likely get exceptions when trying to serialize.

Reply:
- The only usage of method `RoutedClientResultsManager.TryGetInvocationReturnType` is to be called by [`ServiceLifetimeManagerBase.TryGetReturnType` ](https://github.com/Azure/azure-signalr/pull/1687/files#diff-421d2ddd4830ecaab4e6abbccef7751e56cec4eea3e9c0dec81f2a00cc1a1268R348). 
- And `ServiceLifetimeManagerBase.TryGetReturnType` is called by [`DefaultHubDispatcher.DispatchMessageAsync`](https://github.com/dotnet/aspnetcore/blob/main/src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs#L179). The return value `type` is passed into an unused variable `out _`
- After a server receives a `CompletionMessage` corresponding to a routed invocation
  - the server doesn't care about what the return type and `Result` value the invocation owns. 
  - First, [`SetConnectionResultAsync`](https://github.com/Azure/azure-signalr/pull/1687/files#diff-421d2ddd4830ecaab4e6abbccef7751e56cec4eea3e9c0dec81f2a00cc1a1268R335) is called `ServiceLifetimeManagerBase.TryGetReturnType` is called, where the server serializes the `CompletionMessage`, wraps it into `ClientCompletionMessage` and then send `ClientCompletionMessage` out.
- So, The return value `type` of `RoutedClientResultsManager.TryGetInvocationReturnType` is actually useless. 
- So, using `object` here is OK

Is the conclusion above correct?

Reply:
> The only usage of method RoutedClientResultsManager.TryGetInvocationReturnType is to be called by [ServiceLifetimeManagerBase.TryGetReturnType](https://github.com/Azure/azure-signalr/pull/1687/files#diff-421d2ddd4830ecaab4e6abbccef7751e56cec4eea3e9c0dec81f2a00cc1a1268R348).

`ServiceLifetimeManagerBase.TryGetReturnType` is used by `IHubProtocol` when it is parsing the completion message from the client.
https://source.dot.net/#Microsoft.AspNetCore.SignalR.Protocols.Json/Protocol/JsonHubProtocol.cs,233 -> https://source.dot.net/#Microsoft.AspNetCore.SignalR.Protocols.Json/TryGetReturnType.cs,14 -> https://source.dot.net/#Microsoft.AspNetCore.SignalR.Core/Internal/HubConnectionBinder.cs,26
When using `object` as the type then each protocol might handle it differently, for example NewtonsoftJson will create a `JsonObject`. Then when you write the message via `SerializeCompletionMessage` -> https://github.com/Azure/azure-signalr/blob/84f2d69eaf71b497aba216d84d480a1a9738943a/src/Microsoft.Azure.SignalR/Internals/DefaultHubMessageSerializer.cs#L43, each protocol is looped over and tries to serialize the `JsonObject` which will not go well if there are other protocols, e.g. MessagePack.

Reply:
- So it means a server must know the precise return value types of its routed invocations. 
- For a server, information about a routed invocation is all captured from `ClientInvocationMessage` it receives
- Thus, I add a new field `Type` into `ClientInvocationMessage` to let a server knows the real return types of its routed invocations.